Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -344,6 +344,96 @@ theorem of_mulOpposite (h : UniqueProds Gᵐᵒᵖ) : UniqueProds G where
 @[to_additive] instance [h : UniqueProds G] : UniqueProds Gᵐᵒᵖ :=
   of_mulOpposite <| (mulHom_image_iff <| MulEquiv.opOp G).mp h
 
+@[to_additive] private theorem toIsLeftCancelMul [UniqueProds G] : IsLeftCancelMul G where
+  mul_left_cancel a b1 b2 he := by
+    classical
+    have := mem_insert_self b1 {b2}
+    obtain ⟨a, ha, b, hb, hu⟩ := uniqueMul_of_nonempty ⟨a, mem_singleton_self a⟩ ⟨b1, this⟩
+    cases mem_singleton.mp ha
+    simp_rw [mem_insert, mem_singleton] at hb
+    obtain rfl | rfl := hb
+    · exact (hu ha (mem_insert_of_mem <| mem_singleton_self b2) he.symm).2.symm
+    · exact (hu ha this he).2
+
+open MulOpposite in
+@[to_additive] theorem toIsCancelMul [UniqueProds G] : IsCancelMul G where
+  mul_left_cancel := toIsLeftCancelMul.mul_left_cancel
+  mul_right_cancel _ _ _ h :=
+    op_injective <| toIsLeftCancelMul.mul_left_cancel _ _ _ <| unop_injective h
+
+/-! Two theorems in [Andrzej Strojnowski, *A note on u.p. groups*][Strojnowski1980] -/
+
+/-- If unique pairs exist in a group whenever the two `Finset`s are the same,
```

Comment:
This docstring doesn't make much sense outside the context of this file. E.g., "which two finsets?". Would it make sense to mention Storjnowski in the docstring?